drivers: flash: Introduce i.MX RT FlexSPI driver

MaureenHelm opened the pull request:
Introduces a new flash driver for the FlexSPI peripheral on i.MX RT
SoCs. The hardware provides a flexible sequence engine (LUT) that
supports various types of external devices, including serial NOR flash,
serial NAND flash, HyperBus (HyperFlash/HyperRAM), and FPGAs. It
supports up to four connected devices in single/dual/quad/octal modes
and provides memory-mapped read/write access to these devices through
the AHB bus.

The driver implementation consists of a shared controller for each
FlexSPI peripheral instance, and protocol-specific device drivers for
each external device. The controller provides a private interface for
multiple devices to access the FlexSPI peripheral registers. FlexSPI
devices provide the public flash driver interface to applications or
subsystems like storage or flash file systems; they also provide
protocol-specific LUT sequences to the controller.

Currently the only device type supported is QSPI NOR flash, but other
types like HyperFlash will be added later.

XIP is not yet supported, as this requires additional work to relocate
code to RAM and managing interrupts.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
modules: nxp_imx: Add HAS_MCUX_FLEXSPI

Adds a hidden config symbol HAS_MCUX_FLEXSPI selected by NXP SoCs when
the FlexSPI peripheral is present. It will be used as a dependency for a
new FlexSPI flash driver to prevent users from accidentally enabling the
driver on platforms that don't have the necessary hardware.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: arm: Add qspi flash dts node to mimxrt1064_evk

Copies the QSPI flash device tree node from the mimxrt1060_evk to the
mimxrt1064_evk board.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
dts: boards: arm: Rework FlexSPI bindings on i.MX RT boards

Reworks the NXP FlexSPI device tree bindings to configure controller and
device properties needed for an upcoming FlexSPI flash driver.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
drivers: flash: Introduce i.MX RT FlexSPI driver

Introduces a new flash driver for the FlexSPI peripheral on i.MX RT
SoCs. The hardware provides a flexible sequence engine (LUT) that
supports various types of external devices, including serial NOR flash,
serial NAND flash, HyperBus (HyperFlash/HyperRAM), and FPGAs. It
supports up to four connected devices in single/dual/quad/octal modes
and provides memory-mapped read/write access to these devices through
the AHB bus.

The driver implementation consists of a shared controller for each
FlexSPI peripheral instance, and protocol-specific device drivers for
each external device. The controller provides a private interface for
multiple devices to access the FlexSPI peripheral registers. FlexSPI
devices provide the public flash driver interface to applications or
subsystems like storage or flash file systems; they also provide
protocol-specific LUT sequences to the controller.

Currently the only device type supported is QSPI NOR flash, but other
types like HyperFlash will be added later.

XIP is not yet supported, as this requires additional work to relocate
code to RAM and managing interrupts.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
soc: arm: Enable FlexSPI flash driver on i.MX RT family

Enables the FlexSPI flash driver on the i.MX RT SoC family and
configures the peripheral clocks accordingly. We are careful to only
configure the peripheral clocks if we are not executing in place from
the FlexSPI flash.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: arm: Configure FlexSPI QSPI flash on mimxrt1064_evk

Enables the FlexSPI NOR flash driver, configures the FlexSPI pins, and
updates the board documentation accordingly on the mimxrt1064_evk.

Note that this SoC has two FlexSPI instances: one instance has an
in-package QSPI flash used for XIP; the other instance has a board-level
QSPI flash used for storage, not XIP. This patch enables the flash
driver on the non-XIP flash only.

Tested with:
  - samples/subsys/fs/littlefs
  - samples/drivers/flash_shell

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

boards: arm: Rename flexspi_qspi to flexspi_nor for mimxrt1064_evk

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
samples: fs: Extend littlefs sample to mimxrt1064_evk

Adds the mimxrt1064_evk board to the platform_allow list for the
littlefs sample application.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
boards: arm: Clean up HyperFlash dts nodes on mimxrt10{50,60}_evk

Cleans up the HyperFlash device tree nodes on the mimxrt1050_evk and
mimxrt1060_evk_hyperflash boards to be more consistent with other
FlexSPI child nodes.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

arvid-r commented:
This looks great. Thanks!

I'll try to look at the LUT part. We had lots of problems with that initially.

jhqian commented on boards/arm/mimxrt1064_evk/mimxrt1064_evk.dts:
```diff
@@ -141,9 +146,24 @@ arduino_i2c: &lpi2c1 {};
 };
 
+&flexspi2 {
+	xip;
```
It seems this attribute should be linked to CODE_LOCATION.

MaureenHelm replied:
I'd like to get rid of `CODE_LOCATION` and start using a dts chosen node (`zephyr,flash`) instead. I made an attempt to do that in #30065 but it needs work.

MaureenHelm commented on drivers/flash/flash_mcux_flexspi.c:
```diff
@@ -0,0 +1,155 @@
+	static const struct flash_flexspi_config			\
+		flash_flexspi_config_##n = {				\
+		.base = (FLEXSPI_Type *) DT_INST_REG_ADDR(n),		\
+		.ahb_base = (uint8_t *) 0x60000000,			\
```
It isn't currently defined in dts, but it should be.

MaureenHelm replied:
Done. @galak please have a look

pabigot commented:
I have the mimxrt1060_evk, and am unable to get any of the samples (littlefs, spi_flash, jesd216, latter two with code changes to make it compile) to work.  littlefs builds and flashes, but dies in `CLOCK_InitUsb1Pfd` with no further output, not even the boot message.

Reflashing with hello-world works.

Is this board supposed to work, or does it use the flash in a way that's incompatible with this feature?

MaureenHelm commented:
> I have the mimxrt1060_evk, and am unable to get any of the samples (littlefs, spi_flash, jesd216, latter two with code changes to make it compile) to work. littlefs builds and flashes, but dies in `CLOCK_InitUsb1Pfd` with no further output, not even the boot message.
> 
> Reflashing with hello-world works.
> 
> Is this board supposed to work, or does it use the flash in a way that's incompatible with this feature?

No, this won't work on `mimxrt1060_evk` until XIP is supported. I specifically chose `mimxrt1064_evk` because it has two QSPI flashes - one internal to the package that is used for XIP, and one external mounted on the board that I could use exclusively for storage. `mimxrt1060_evk` only has the latter QSPI flash, so it has to be used for both code and storage, and thus needs XIP support in the flash driver.

MaureenHelm commented:
I pushed a new version with the following changes:
- Renamed the driver and dts compatibles from `qspi` to `nor`. e.g., `flash_mcux_flexspi_nor.c` and `nxp,imx-flexspi-nor`. I want to add octal support in a future PR, and I think it will make more sense to extend this driver rather than create a new driver.
- Removed unused LUT sequences. I tried to do this in my original version, but it didn't work at runtime. Turns out I needed to change the init function to update the LUT *after* setting the flash config, not before.
- Included `spi_nor.h` to reuse some command macros like `SPI_NOR_CMD_RDID` and `SPI_NOR_CMD_RDSR`

@galak Will you please review the dts bindings?
@mmahadevan108 @arvid-r Will you please review the driver implementation?

pabigot commented on dts/bindings/mtd/nxp,imx-flexspi-device.yaml:
```diff
@@ -0,0 +1,85 @@
+    cs-interval:
+      type: int
+      required: false
+      default: 2
```
Where do these defaults come from?  https://github.com/zephyrproject-rtos/zephyr/blob/2d215484cc352152e9686477fead841c13ebb249/dts/binding-template.yaml#L244

pabigot commented on drivers/flash/flash_mcux_flexspi.c:
```diff
@@ -0,0 +1,155 @@
+	bool ahb_read_addr_opt;
+	bool combination_mode;
+	bool xip;
+	flexspi_read_sample_clock_t rx_sample_clock;
```
As an enum this could be `sizeof(int)`; some alignment costs might be avoided by moving it above the `bool` fields.

pabigot reviewed (commented):
A couple minor comments.  I can't test this and don't know the internals well enough to approve based on reading alone.

I think it would make sense to be more clear in the devicetree files for boards like mimxrt1060_evk that the flexspi-nor node doesn't actually work now.  Perhaps by setting `status = "disabled";` with a comment explaining why.

arvid-r commented:
> * Removed unused LUT sequences. I tried to do this in my original version, but it didn't work at runtime. Turns out I needed to change the init function to update the LUT _after_ setting the flash config, not before.
Why is that? Is the LUT cleared by setting the flash config?

arvid-r commented:
Looks good in general, but this is rather complex. I would like to try it on imx1020. Is that supposed to work if I run directly from SRAM? I have always used XIP, so I don't know if that is possible or how it can be done.

galak commented on dts/bindings/spi/nxp,imx-flexspi.yaml:
```diff
@@ -13,3 +13,50 @@ properties:
+      type: int
+      required: true
+      description: |
+        AHB address for read/write access to serial flash memory.
```
How is this address different from the MMIO region that the flash controller specifies?

MaureenHelm replied:
The MMIO region specified by the flash controller contains the controller registers, this is the flash itself memory mapped into the AHB address space.

galak commented on dts/bindings/spi/nxp,imx-flexspi.yaml:
```diff
@@ -13,3 +13,50 @@ properties:
+      description: |
+        Combine port A and port B data pins to support octal mode access.
+
+    xip:
```
What do we need this vs CONFIG_XIP?

MaureenHelm replied:
So we can tell which flash we're using for XIP

galak replied:
I seems like we should in theory, but would need a bit more of a use case as your question implies a scenario with multiple flash's.

I assume in the case we'd be able to tell if CONFIG_XIP + looking at `zephyr,flash-controller` or looking if `zephyr,flash` is inside the MMIO region for the specific controller instance.

MaureenHelm replied:
The `mimxrt1064_evk` board is a scenario with multiple flashes - one internal to the SoC package connected to the FLEXSPI2 controller, and a second one mounted on the board connected to the FLEXSPI1 controller. I introduced this property to avoid reinitializing FLEXSPI2 while we're trying to execute from the internal flash (it was already initialized by the boot ROM); otherwise code execution goes off in the weeds.

As @pabigot noted, an alternative approach would be to set `status = "disabled"` on FLEXSPI2, however we are using that node to generate `FLASH_SIZE` and `FLASH_BASE_ADDRESS`. It's effectively the `zephyr,flash` chosen node.

galak replied:
So looks like we can make flexspi2 as `status="disabled"`.

So lets go with that path and remove the `xip` property.

galak commented on dts/bindings/spi/nxp,imx-flexspi.yaml:
```diff
@@ -13,3 +13,50 @@ properties:
+      type: boolean
+      required: false
+      description: |
+        Enable AHB bufferable write access.
```
For these, it might be good to add to the `description` something like see details in the `AHBCR register in UM for more details`

galak reviewed (changes requested):
Marking request changes for now as I try to understand some DTS related questions:

* What properties are you inheriting from the jedec nor binding?  Seems like just the `size`
* Why do we need different compats for hyperflash v nor as the bindings are identical currently
* Why the defaults (see @pabigot question) on this.
* other inline comments/queries.

MaureenHelm commented:
> I think it would make sense to be more clear in the devicetree files for boards like mimxrt1060_evk that the flexspi-nor node doesn't actually work now. Perhaps by setting `status = "disabled";` with a comment explaining why.

We can't disable the node because it's used to generate FLASH_SIZE and FLASH_BASE_ADDRESS for XIP.

> Looks good in general, but this is rather complex. I would like to try it on imx1020. Is that supposed to work if I run directly from SRAM? I have always used XIP, so I don't know if that is possible or how it can be done.

I've been working to extend this to the XIP flash on RT1064, but it's not quite functional yet. I can relocate the driver to RAM with #31132, but unless we relocate the entire app to RAM, we have to disable interrupts for a very long time when we write or erase the flash.

> * What properties are you inheriting from the jedec nor binding?  Seems like just the `size`

size and spi-device properties

> * Why do we need different compats for hyperflash v nor as the bindings are identical currently

Because in the future we'll introduce another driver for hyperflash, which needs a different LUT and driver logic than the nor driver.

> * Why the defaults (see @pabigot question) on this.

I used the values needed for the internal QSPI flash on RT1064.

pabigot commented:
> I used the values needed for the internal QSPI flash on RT1064.

If those are not going to be correct for other QSPI flash devices then they need to be explicit in the rt1064 and other board devicetree files, and required rather than defaulted in the devicetree binding specification.

GrixaYrev commented on drivers/flash/flash_mcux_flexspi.c:
```diff
@@ -0,0 +1,155 @@
+		.combination_mode = DT_INST_PROP(n, combination_mode),	\
+		.xip = DT_INST_PROP(n, xip),				\
+		.rx_sample_clock =					\
+			kFLEXSPI_ReadSampleClkLoopbackFromDqsPad,	\
```
Option `kFLEXSPI_ReadSampleClkLoopbackFromDqsPad` require pin DQS, but this pin can be not available on some boards. So it is more flexible to use DTS for setup this option.

MaureenHelm replied:
I agree

GrixaYrev commented on soc/arm/nxp_imx/rt/soc.c:
```diff
@@ -213,6 +213,13 @@ static ALWAYS_INLINE void clock_init(void)
+	CLOCK_DisableClock(kCLOCK_FlexSpi);
+	CLOCK_InitUsb1Pfd(kCLOCK_Pfd0, 24);
+	CLOCK_SetMux(kCLOCK_FlexspiMux, 3);
+	CLOCK_SetDiv(kCLOCK_FlexspiDiv, 2);
```
I have tested the FlexSPI driver on board [VisionSOM-RT](https://wiki.somlabs.com/index.php/VisionSOM-RT) with i.MX RT1052. When FlexSPI root clock is equal 120 MHz, amplitude of signal FlexSPI_CLK is too small. I attach two oscilloscope screenshot, with `FlexSpiDiv==2` and `FlexSpiDiv==3` (120 MHz and 90 MHz). It is depended on board capacity, so i suggest to use DTS for setup value of this divider.
![Div2](https://user-images.githubusercontent.com/73971382/104431189-a8c7fa00-5598-11eb-87a8-c849eca7bff0.png)![Div3](https://user-images.githubusercontent.com/73971382/104431191-a9609080-5598-11eb-847f-3622395a6f60.png)

GrixaYrev commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -0,0 +1,515 @@
+	},
+
+	[ERASE_SECTOR] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR,       kFLEXSPI_1PAD, 0xD7,
```
Why is command `D7h` used here, not `20h` (`SPI_NOR_CMD_SE`)? W25Q128JV as well as IS25WP064A support command `20h`, but W25Q128JV not support `D7h`. I found command `20h` in JESD251A (table 5), but i cannot understand is this command standard.

pabigot replied:
This is an interesting point.  The command might be defaultable, but for portability it should come from the JESD216 BFP DW8 and DW9.  Is it possible to interact with the device using a default LUT first, to extract the SFDP data, then program a LUT that is correct for the discovered device?

MaureenHelm replied:
The LUT is derived from the MCUXpresso SDK example `evkmimxrt1064_flexspi_nor_polling_transfer`. I don't know why `0xD7` was used there instead of `0x20`, but experimentally it seems that the standard command also works with IS25WP064.

arvid-r replied:
How can different flash chips be supported in the future? I guess the LUT might be subject to change if another chip is used, but now it is hardcoded in the driver.

GrixaYrev replied:
For exapmle, board [VisionSOM-RT](https://wiki.somlabs.com/index.php/VisionSOM-RT) have flash [W25Q128JV](https://www.winbond.com/resource-files/w25q128jv%20revf%2003272018%20plus.pdf), which do not support `D7h`. Also [S25FL128L](https://www.cypress.com/file/316171/download) by CYPRESS and [MT25QL128](https://www.micron.com/products/nor-flash/serial-nor-flash/part-catalog?Value=128Mb&Id={826DD0CE-31A2-43CD-AA36-E65D0DB21D31}) by Micron support command `20h`. So, it is universal to use command `20h`.

pabigot replied:
One of the main purposes of adding SFDP support in Zephyr is to allow the driver to get the information (like erase command) from the flash chip directly, and configure itself.  I do think doing that would handle the concerns in this thread, but other than changing the command to the standard 20h from the unusual D7h I don't think any of that has to be done now.

arvid-r commented on drivers/flash/flash_mcux_flexspi_nor.c:
```diff
@@ -0,0 +1,515 @@
+	READ_FAST_QUAD_OUTPUT,
+	PAGE_PROGRAM_QUAD_INPUT,
+	ENTER_QPI,
+};
```
Is the order of these commands important? Does it have to correspond to the look-up table index in Table 8-15 in the iMX RT1020 Reference Manual for example, for XIP to work in the future. Maybe I misunderstood how this works. It is a very flexible system, for good and bad.

MaureenHelm replied:
Based on the footnote after the table, I think we can change them:
```
1. All the pre-defined LUT indexes are only applicable to
boot stage. User application can use the whole 16 LUT
entries freely based on their requirement.
```

I ordered them this way because I think we may eventually want to conditionalize quad and octal support. Something like this:
```c
enum {
        /* SPI instructions */
        READ_ID,
        READ_STATUS_REG,
        WRITE_STATUS_REG,
        WRITE_ENABLE,
        PAGE_PROGRAM,
        ERASE_SECTOR,
        ERASE_CHIP,

#ifdef CONFIG_FLASH_FLEXSPI_NOR_QUAD
        /* Quad SPI instructions */
        READ_FAST_QUAD_OUTPUT,
        PAGE_PROGRAM_QUAD_INPUT,
        ENTER_QPI,
#endif

#ifdef CONFIG_FLASH_FLEXSPI_NOR_OCTAL
        /* Octal SPI instructions */
        READ_FAST_OCTAL_OUTPUT,
        PAGE_PROGRAM_OCTAL_INPUT,
        ENTER_OPI,
#endif
};

```

MaureenHelm commented:
I pushed a new version with the following changes:
- Expanded dts property descriptions to include references to their associated hardware register fields
- Reviewed and updated dts property defaults to reflect hardware register reset values. There are a few exceptions for boolean properties that are default enabled in hardware, but default disabled in dts. I believe it's not possible to unset a boolean dts property if it was already set by default.
- Added an `rx_clock_source` property instead of hard-coding the value in the driver
- Removed the `ahb-base` property and used the second MMIO reg property instead (`DT_INST_REG_ADDR_BY_IDX(n, 1)`)
- Replaced the chip-specific sector erase command with the standard `SPI_NOR_CMD_SE` command

I purposely did not rebase to make it easier to see what changed from the previous version.

pabigot commented:
> I believe it's not possible to unset a boolean dts property if it was already set by default.

It is; in an overlay add:

```
&node {
    /delete-property/ bool-prop;
};
```

MaureenHelm commented:
Rebased to resolve conflicts after merging #30864 and to pick up #31042.

@pabigot please check that I've integrated the jesd216 binding correctly in `dts/bindings/mtd/nxp,imx-flexspi-device.yaml`

`scripts/dts/gen_defines.py` wouldn't let me set boolean properties as default true, so I set them in the SoC level `dts/arm/nxp/nxp_rt.dtsi`

mbolivar-nordic commented:
> `scripts/dts/gen_defines.py` wouldn't let me set boolean properties as default true, so I set them in the SoC level `dts/arm/nxp/nxp_rt.dtsi`

I think that's an edtlib bug; I'm working on a patch.

mbolivar-nordic commented:
> I think that's an edtlib bug; I'm working on a patch.

https://github.com/zephyrproject-rtos/zephyr/pull/31366

mbolivar-nordic commented on drivers/flash/flash_mcux_flexspi.c:
```diff
@@ -0,0 +1,154 @@
+
+	FLEXSPI_GetDefaultConfig(&flexspi_config);
+
+	flexspi_config.ahbConfig.enableAHBBufferable = config->ahb_bufferable;
```
Since we can't "fix" this sort of thing in edtlib, an alternative is to invert the devicetree property meaning. Something like `no-ahb-bufferable` in devicetree, and `.ahb_bufferable = !DT_INST_PROP(..., no_ahb_bufferable)` when initializing `config`.

Feel free to disregard if you're happier with setting this in .dtsi.

MaureenHelm replied:
I think it would be confusing to tie a property to hardware register field but invert the logic, so keeping this in the dtsi. Thanks for your help

pabigot commented on boards/arm/mimxrt1010_evk/mimxrt1010_evk.dts:
```diff
@@ -46,7 +46,7 @@ arduino_serial: &lpuart1 {};
 	reg = <0x402a8000 0x4000>, <0x60000000 DT_SIZE_M(16)>;
 	at25sf128a: at25sf128a@0 {
-		compatible = "adesto,at25sf128a", "jedec,spi-nor";
+		compatible = "nxp,imx-flexspi-nor";
```
It'd be fine to have this be:

    compatible = "adesto,at25sf128a", "nxp,imx-flexspi-nor";

Then if somebody has an Adesto-specific driver it'd be picked up.  It also makes more clear exactly what the flash chip is.  Not required, but suggested.

The same applies to all the other flash device nodes.

MaureenHelm replied:
I like how this makes the flash chip more clear, but for it to work I think we'd need a compatible something like `nxp,imx-flexspi-nor-at25sf128a`. Because an Adesto-specific driver would be different for each of the various SoC QSPI peripherals.

pabigot replied:
I'm not sure that's necessarily true.  I'm thinking this would be useful in somebody's out-of-tree application where they didn't care that the underlying peripheral isn't represented in the compatible name, but their needs are such that their dedicated custom driver is preferred over the stock Zephyr one.

The primary purpose of the compatible is to accurately describe the device that's present in hardware, not the driver implementation to talk to that device.  Kconfig figures out the best device.  QSPI bus devices are perhaps an extreme case of this, but could be supported if we used `jedec,qspi-nor` and have the vendor-specific drivers all support that compatible, if it weren't for the fact the vendor drivers tend to need vendor-specific devicetree properties.

pabigot commented on dts/bindings/mtd/nxp,imx-flexspi-device.yaml:
```diff
@@ -0,0 +1,101 @@
+
+compatible: "nxp,imx-flexspi-device"
+
+include: [spi-device.yaml, "jedec,jesd216.yaml"]
```
This looks right to me.

MaureenHelm replied:
Thanks for checking

galak reviewed (commented):
There is a bug in mimxrt1064_evk in how `zephyr,flash-controller` is set.  It should technically not be set, based on how we define `zephyr,flash-controller` as the controller node for where `zephyr,flash` is set.  As we don't set `zephyr,flash` we shouldn't be setting this property.

This seems to be used by various tests & samples because of assumption that the flash controller we'd use for a test/sample would be the same as the one that Zephyr code is running out of.

MaureenHelm committed:
drivers: flash: Fix flash shell to work without dts chosen node

The flash shell can work with any flash driver instance, not just the
one chosen by zephyr,flash-controller. It's helpful for the flash shell
to use this instance by default, but not required. We can switch
instances at runtime with the "flash set_device" command.

Fix the flash shell so it can build when there isn't a chosen
zephyr,flash-controller available.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm committed:
samples: flash_shell: Run on boards without zephyr,flash-controller

The flash shell can now build when there isn't a chosen
zephyr,flash-controller available, so we can simplify the filter to
apply this sample to more boards.

Two STM32 H7 board configurations are excluded because the flash driver
isn't supported yet on the M4 core.

Signed-off-by: Maureen Helm <maureen.helm@nxp.com>

MaureenHelm commented:
> There is a bug in mimxrt1064_evk in how `zephyr,flash-controller` is set. It should technically not be set, based on how we define `zephyr,flash-controller` as the controller node for where `zephyr,flash` is set. As we don't set `zephyr,flash` we shouldn't be setting this property.

Ok, fixed. Please have another look. I also removed the xip property and set `status = "disabled"` instead.

galak commented on soc/arm/nxp_imx/rt/soc.c:
```diff
@@ -213,6 +213,13 @@ static ALWAYS_INLINE void clock_init(void)
 	CLOCK_SetMux(kCLOCK_CanMux, 2); /* Set Can clock source. */
 #endif
 
+#if defined(CONFIG_FLASH_MCUX_FLEXSPI) && !DT_PROP(DT_NODELABEL(flexspi), xip)
```
NIT: you've removed the `xip` property so you can drop the test for it here.

MaureenHelm replied:
Good catch, fixed.

galak reviewed (commented):
Minor issue w/XIP prop in soc.c that can be removed.
